File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.java
class Solution {
    public int lengthOfLongestSubstring(String s) {
        
        int maxlen = 0;
        int i=0, j=0;
        
        HashSet<Character> set = new HashSet();
        
        while(j < s.length()){
            if(!set.contains(s.charAt(j))){
                set.add(s.charAt(j++));
                maxlen = Math.max(maxlen, set.size());
            }else{
                set.remove(s.charAt(i++));
            }
        }
        return maxlen;
    }
}

File: 3-longest-substring-without-repeating-characters/3-longest-substring-without-repeating-characters.cpp
class Solution {
public:
    int lengthOfLongestSubstring(string s) {
        
        if(s.length() == 0)
            return 0;
        
        unordered_map<char, int> map;
        int maxlen=0;
        
        int i=0, j=0;
        
        for(i=0; i<s.length(); i++){
            if(map.find(s[i]) != map.end()){
                j = max(j,map[s[i]]+1);
            }
            
            map[s[i]] = i;
            maxlen = max(maxlen, i-j+1);
        }
        
        return maxlen;
        
        
    }
};